Fixes Issues 2865, 2866, 2870, 2871 in CriticalPath AsyncCSS and DOM Module

hellofromtonya opened the pull request:
Closes #2865 
Closes #2866 
Closes #2870
Closes #2871 

Closes #2885

Adds the updated DOM Module and its test cases to fix the following issues:

- [x] Bugfix: character encoding 
     - [x] improves encoding to handle non-Latin characters (different languages added to the test cases to validate)
     - [x] Auto detect character encoding - see [DOM Module issue 21](https://github.com/wp-media/module-dom/issues/21)
     - [x] Preserve HTML entities after processing - see [DOM Module Issue 22](https://github.com/wp-media/module-dom/issues/22)
- [x] Bugfix: Adds guard clauses to prevent PHP notice when an empty string is passed into `HTMLDocument::from_html`, `HTMLDocument::from_fragment`, and `HTMLDocument:loadHtml`.
- [x] Bugfix: Adds guard clause to `HTMLDocument::query` to ensure XPath is set up
- [x] Adds test cases to validate that the `type` attribute is not being stripped from `<script>` element
- [x] Bugfix: Preserves HTML templates inside `<script>` elements
- [x] Preserves malformed and invalid content inside of the `<head></head>` 
- [x] Preserves`<noscript>` in `<head>` when libxml is < 2.8.0
- [x] Attempts to fix/normalize the basic HTML structure of
```html
<!DOCTYPE html>
<html>
<head>
</head>
<body>
</body>
</html>
```
- [x] Transforms self-closing tags to HTML5 format, e.g. `<br/>`, `<br />` , `<br         />` etc are transformed into `<br>`
- [x] Regression: include media attribute when loading CSS files async Issue #2885

Credit props to @engahmeds3ed for identifying the encoding issue with non-Latin characters 🥇

hellofromtonya committed:
Copies latest HTMLDocument file from the DOM Module.

hellofromtonya committed:
Adds new file and test from the DOM Module.;

hellofromtonya committed:
Mocks get_option in AsyncCSS unit tests.

hellofromtonya committed:
Adds guard clause to bail out of AsyncCSS::from_html.

hellofromtonya commented:
This PR is ready for @wp-media/qa 

Things to check in this one:

1. Add non-Latin characters or language to a web page and check that the characters remain intact. For example, [check out this test data in the fixture](https://secure.helpscout.net/conversation/1219841185/179108).
2. Testing an empty string's impact can really only be done by [modifying the code here](https://github.com/wp-media/wp-rocket/blob/updates/dom-module-for-fixes/inc/Engine/CriticalPath/CriticalCSSSubscriber.php#L591) to pass an `''` empty string instead of `$html`.

Davidacu reviewed (approved):
Testing Environment:
http://mega.wp-rocket.me/qa/

1. I can confirm that non-Latin characters are being handled properly now.
Before in 3.6.2 -> https://i.imgur.com/yTk9ZJW.jpg
Now -> https://i.imgur.com/2NJmtH4.jpg

2. I manually modified the [code here](https://github.com/wp-media/wp-rocket/blob/updates/dom-module-for-fixes/inc/Engine/CriticalPath/CriticalCSSSubscriber.php#L591) to pass an empty string and did not generate any PHP warnings on Critical CSS regeneration.

3. Confirmed the `<link>` elements inside `<noscript>` at the bottom are valid elements, chrome does show them as a string if JS is enabled, but if JS is disabled it does show as elements.

hellofromtonya committed:
Adds test data for ll noscript.

hellofromtonya committed:
Adds more tests for ll script/noscript.

hellofromtonya committed:
Adds ll image to test cases.

hellofromtonya committed:
Fixes stripping of HTML closing tags inside of a <script> template.

How?

1. Adds LIBXML_SCHEMA_CREATE as an option when loadHTML() is invoked.
2. Changes the iterator to an array when changing a live Node List.

Adds test cases.

hellofromtonya committed:
Adds the 3.6.2.1 DOM Module.

hellofromtonya committed:
Updates AsyncCSS query, which is now an array coming back.

hellofromtonya committed:
Fixes test content for new rules.

hellofromtonya committed:
Fixes the test.

hellofromtonya commented:
This PR has been updated. Let's review what's new and changed since the last QA and code review:

Websites are filled with invalid and malformed HTML/content. Before DOM Module, Rocket didn't care if the HTML was well-formed or not. With `DOMDocument`, the DOM Module needs to do more to be flexible to work with invalid and malformed HTML/content.

The new DOM Module (bundled in this PR) is now more flexible. 

- Preserves invalid and/or malformed HTML/content in the `<head>`. Issue #2871
- Adds missing basic HTML structure tags when missing: Issue #2871
    - Adds `<DOCTYPE>`, `<html>`, `<head>`, `</head>`, `<body>`, `</body>`, `</html>`
    - However, if the `</head>` and `<body>` tags are missing, we bail out. Why? We can't guess where those should be inserted. If both are missing, then AsyncCSS (Optimized CSS Delivery) is not applied to the web page.
- Preserves closing tags contained with `<script>`, i.e. for JS-powered templates to insert HTML.
- Preserves all `<noscript>` elements in the `<head>`, even if the hosting Libxml version is < 2.8.0. Issue #2871
- Preserves the charset encoding for non-Latin characters, i.e. Issue #2866
- And retains the fixes for Issues #2865 and #2870

hellofromtonya commented:
@wp-media/qa This PR is ready for re-QA.

Tabrisrp commented on inc/Engine/DOM/Transformer/Normalizer.php:
```diff
@@ -0,0 +1,296 @@
+	/**
+	 * Normalize the document structure.
+	 *
+	 * This makes sure the document adheres to the general structure that AMP requires:
```
```suggestion
	 * This makes sure the document adheres to the general structure that WP Rocket requires:
```

hellofromtonya replied:
Resolved https://github.com/wp-media/wp-rocket/pull/2868/commits/b769cd142e8545dbbcad4d476072f5993763db55

hellofromtonya committed:
Updates normalize structure docblock.

hellofromtonya committed:
Inject encoding into Transformer.

hellofromtonya committed:
Makes string manipulations multibyte safe.

hellofromtonya committed:
Adds and improves test cases.

hellofromtonya committed:
Bail out of AsyncCSS::from_html if HTMLDocument::from_html does not return object.

There are several conditions which can cause this to happen:

- Empty HTML given
- Missing </head> and <body> tags (too malformed)
- Can't determine the charset
- Server set up or library problem

hellofromtonya committed:
Prevents applying async css to links inside of a <noscript>.

hellofromtonya committed:
Improves comments.

hellofromtonya committed:
Adds newly updated DOM Module with new encoding approach.

hellofromtonya committed:
Updates tests.

hellofromtonya committed:
Fixes issue 2885 - keep original media on preload link.

hellofromtonya committed:
Fix phpcs in AsyncCSS file.

hellofromtonya commented:
Update on this PR:

After testing on multiple customer sites, we found 1:30 had missing `</div>` tags. These types of unclosed tags break sites as `DOMDocument` is unable to identify the exact `<div>` that was not closed. Why? `<div>` can be nested.

Knowing that this is a breaking issue, we decided to add a new feature to the DOM Module: apply only changes made to the DOM to the actual HTML.

How is this different?

Instead of converting the entire DOM back into a HTML string, we will only apply mapped changes made to the string. In this way, invalid tags, malformed HTML, and/or unclosed tags are not edited or changed, i.e. the original HTML for these remains.

I'm closing this PR and will open a new PR to capture the new enhancements.
